### PR TITLE
optimize cssColor check regex

### PR DESCRIPTION
**Pull request by RedSparr0w**
This PR aims to optimize the [regex](https://github.com/badges/shields/blob/master/lib/make-badge.js#L92) used to check hex colors, which is run [2 times](https://github.com/badges/shields/blob/master/lib/make-badge.js#L130-L131) for every badge served.

From testing this regex should be ~10% faster.
_(it's not as if this regex check is using many resources but every bit helps)_

Testing done using [benchmark (npm)](https://www.npmjs.com/package/benchmark).

```javascript
var colors = ["#4c1","#4C1","#abc123","#ABC123","red","#123red","#red"];
var isCSS1 = /^#([0-9a-fA-F]{3}){1,2}$/; // This PR
var isCSS2 = /^#([0-9a-f]{3}){1,2}$/i; // This PR (case insensitive)
var isCSS3 = /^#([0-9a-fA-F]{6}|[0-9a-fA-F]{3})$/; // Current
var isCSS4 = /^#([0-9a-f]{6}|[0-9a-f]{3})$/i; // Current (case insensitive)
var isCSS5 = /^#([0-9a-fA-F]{3}|[0-9a-fA-F]{6})$/; // Current reversed (case insensitive)
var isCSS6 = /^#([0-9a-f]{3}|[0-9a-f]{6})$/i; // Current reversed (case insensitive)
```

The odd numbered test were run asynchronously:
image

Also did some test in browser with [jsben.ch](http://jsben.ch/3mf2N):
image

### CODE REVIEW

File: lib/make-badge.js [hunk 1]
```diff
@@ -89,7 +89,7 @@ function compressSvg(string, callback) {
   svgo.optimize(string, callback);
 }
 
-const cssColor = /^#([0-9a-fA-F]{6}|[0-9a-fA-F]{3})$/;
+const cssColor = /^#([0-9a-f]{3}){1,2}$/i;
```
Comment (paulmelnikow) [inline lib/make-badge.js]: Could you leave a comment in the code that this is an optimized regex, with a link to this PR? Wouldn't want anyone to tweak it and undo this work!

**Comment by platan** [comment]
@RedSparr0w This looks great! Do you still have a fragment of code which you run? Can you paste it here? I would like to run this benchmark on my PC. Which version of node did you use?

File: lib/make-badge.js (hunk 1 above)
Comment (RedSparr0w) [reply lib/make-badge.js]: Good idea, will add that.

**Comment by RedSparr0w** [comment]
@platan 
<details>
<summary>I don't have the original one i used but here's a similar one with a couple more test added:</summary>

_Excuse the naming, wasn't sure what to call them all :smile:_

OS: `Windows Server 2008`
Node: `v8.9.3`
image
```javascript
var Benchmark = require('benchmark');

var suite = new Benchmark.Suite;

var colors = ["#4c1","#4C1","#abc123","#ABC123","red","#123red","#red"];

var isCSS1 = /^#([0-9a-fA-F]{3}){1,2}$/; //pr1
var isCSS2 = /^#([0-9a-f]{3}){1,2}$/i; //pr1 (i)
var isCSS3 = /^#([a-fA-F0-9]{3}){1,2}$/; //pr2
var isCSS4 = /^#([a-f0-9]{3}){1,2}$/i; //pr2 (i)
var isCSS5 = /^#([\da-fA-F]{3}){1,2}$/; //pr3
var isCSS6 = /^#([\da-f]{3}){1,2}$/i; //pr3 (i)
var isCSS7 = /^#([a-fA-F\d]{3}){1,2}$/; //pr4
var isCSS8 = /^#([a-f\d]{3}){1,2}$/i; //pr4 (i)
var isCSS9 = /^#([0-9a-fA-F]{6}|[0-9a-fA-F]{3})$/; // c1
var isCSS10 = /^#([0-9a-f]{6}|[0-9a-f]{3})$/i; // c1 (i)
var isCSS11 = /^#([0-9a-fA-F]{3}|[0-9a-fA-F]{6})$/; // c2
var isCSS12 = /^#([0-9a-f]{3}|[0-9a-f]{6})$/i; // c2 (i)
var isCSS13 = /^#([\da-fA-F]{6}|[\da-fA-F]{3})$/; // c3
var isCSS14 = /^#([\da-f]{6}|[\da-f]{3})$/i; // c3 (i)
var isCSS15 = /^#([\da-fA-F]{3}|[\da-fA-F]{6})$/; // c4
var isCSS16 = /^#([\da-f]{3}|[\da-f]{6})$/i; // c4 (i)

// add tests 
suite
.add('this PR_1', function() {
  colors.forEach((color) => {isCSS1.test(color);});
})
.add('this PR_1 (case insensitive)', function() {
  colors.forEach((color) => {isCSS2.test(color);});
})
.add('this PR_2', function() {
  colors.forEach((color) => {isCSS3.test(color);});
})
.add('this PR_2 (case insensitive)', function() {
  colors.forEach((color) => {isCSS4.test(color);});
})
.add('this PR_3', function() {
  colors.forEach((color) => {isCSS5.test(color);});
})
.add('this PR_3 (case insensitive)', function() {
  colors.forEach((color) => {isCSS6.test(color);});
})
.add('this PR_4', function() {
  colors.forEach((color) => {isCSS7.test(color);});
})
.add('this PR_4 (case insensitive)', function() {
  colors.forEach((color) => {isCSS8.test(color);});
})
.add('current_1', function() {
  colors.forEach((color) => {isCSS9.test(color);});
})
.add('current_1 (case insensitive)', function() {
  colors.forEach((color) => {isCSS10.test(color);});
})
.add('current_2', function() {
  colors.forEach((color) => {isCSS11.test(color);});
})
.add('current_2 (case insensitive)', function() {
  colors.forEach((color) => {isCSS12.test(color);});
})
.add('current_3', function() {
  colors.forEach((color) => {isCSS13.test(color);});
})
.add('current_3 (case insensitive)', function() {
  colors.forEach((color) => {isCSS14.test(color);});
})
.add('current_4', function() {
  colors.forEach((color) => {isCSS15.test(color);});
})
.add('current_4 (case insensitive)', function() {
  colors.forEach((color) => {isCSS16.test(color);});
})
.on('cycle', function(event) {
  console.log(String(event.target));
})
.on('complete', function() {
  console.log('-------------------------------------------------\nFastest is ' + this.filter('fastest').map('name') + '\n');
})
.run({ 'async': true })
.run({ 'async': false });
```

I noticed that the case insensitive option wasn't always faster, but it usually had less deviance.
</details>

**Comment by RedSparr0w** [comment]
>By the way, do we not have any tests for this regex and functionality?

@PyvesB Doesn't look like there is any test, Not entirely sure how to add one either.
Would something like this work:
```javascript
  describe('HEX color regex', function () {
    it('should check if given valid css hex color', function () {
      const testColorsTrue = ["#4c1","#4C1","#abc123","#ABC123"];
      const testColorsFalse = ["red","#123red","#red"];
      testColorsTrue.forEach((color) => {
        expect(cssColor.test(color)).to.be.true;
      });
      testColorsFalse.forEach((color) => {
        expect(cssColor.test(color)).to.be.false;
      });
    });
  });
```

**Comment by PyvesB** [comment]
@RedSparr0w would it be possible to write a similar test by going through the `makeBadge` function instead? Not sure whether it would be better, but I would personally be inclined to try to avoid exporting variables just for testing purposes if possible. ^^

**Comment by RedSparr0w** [comment]
@PyvesB No worries, updated the test, could you give them another look.
I feel like there is a better way to check it equals what it should that i may be missing?
```javascript
      colorsTrue.forEach((color) => {
        //option 1:
        expect(JSON.parse(makeBadge({ text: ['name', 'Bob'], colorB: color, format: 'json', template: '_shields_test' })).colorB).to.be(color);
        //option 2:
        expect(makeBadge({ text: ['name', 'Bob'], colorB: color, format: 'json', template: '_shields_test' })).to.include(color);
        //option 3:???
      });
```

**Comment by PyvesB** [comment]
I can't think of anything better, option 1, even though a bit verbose, seems to do the job just fine. 😉

**Comment by platan** [comment]
My results for https://github.com/badges/shields/pull/1602#issuecomment-376685193. 
Node: 8.9.3
Operating System: Ubuntu 16.04.4 LTS
Kernel: Linux 4.13.0-37-generic
Architecture: x86-64
CPU: Intel® Core™ i7-4600U CPU @ 2.10GHz × 4 

Run 1:
```
this PR_1 x 2,591,309 ops/sec ±0.95% (92 runs sampled)
this PR_1 (case insensitive) x 2,541,261 ops/sec ±0.63% (91 runs sampled)
this PR_2 x 2,661,923 ops/sec ±0.74% (94 runs sampled)
this PR_2 (case insensitive) x 2,654,517 ops/sec ±0.47% (92 runs sampled)
this PR_3 x 2,664,610 ops/sec ±0.70% (93 runs sampled)
this PR_3 (case insensitive) x 2,669,557 ops/sec ±0.34% (95 runs sampled)
this PR_4 x 2,613,314 ops/sec ±0.70% (95 runs sampled)
this PR_4 (case insensitive) x 2,684,481 ops/sec ±0.41% (93 runs sampled)
current_1 x 2,316,868 ops/sec ±0.87% (89 runs sampled)
current_1 (case insensitive) x 2,335,650 ops/sec ±0.38% (91 runs sampled)
current_2 x 2,435,518 ops/sec ±0.79% (91 runs sampled)
current_2 (case insensitive) x 2,408,370 ops/sec ±0.47% (90 runs sampled)
current_3 x 2,357,430 ops/sec ±1.08% (94 runs sampled)
current_3 (case insensitive) x 2,377,870 ops/sec ±0.36% (94 runs sampled)
current_4 x 2,416,110 ops/sec ±1.36% (89 runs sampled)
current_4 (case insensitive) x 2,486,850 ops/sec ±0.33% (91 runs sampled)
-------------------------------------------------
Fastest is this PR_4 (case insensitive)

this PR_1 x 2,589,533 ops/sec ±0.40% (86 runs sampled)
this PR_1 (case insensitive) x 2,582,347 ops/sec ±0.33% (91 runs sampled)
this PR_2 x 2,638,306 ops/sec ±1.34% (88 runs sampled)
this PR_2 (case insensitive) x 2,602,063 ops/sec ±0.44% (89 runs sampled)
this PR_3 x 2,659,625 ops/sec ±0.31% (86 runs sampled)
this PR_3 (case insensitive) x 2,690,126 ops/sec ±0.36% (90 runs sampled)
this PR_4 x 2,645,887 ops/sec ±0.36% (89 runs sampled)
this PR_4 (case insensitive) x 2,640,892 ops/sec ±0.33% (93 runs sampled)
current_1 x 2,327,955 ops/sec ±0.42% (91 runs sampled)
current_1 (case insensitive) x 2,326,746 ops/sec ±0.39% (90 runs sampled)
current_2 x 2,445,628 ops/sec ±0.36% (90 runs sampled)
current_2 (case insensitive) x 2,415,592 ops/sec ±0.34% (93 runs sampled)
current_3 x 2,369,587 ops/sec ±0.35% (93 runs sampled)
current_3 (case insensitive) x 2,349,398 ops/sec ±0.43% (93 runs sampled)
current_4 x 2,426,629 ops/sec ±0.39% (93 runs sampled)
current_4 (case insensitive) x 2,521,225 ops/sec ±0.46% (93 runs sampled)
-------------------------------------------------
Fastest is this PR_3 (case insensitive)
```
Run 2:
```
this PR_1 x 2,595,489 ops/sec ±0.75% (92 runs sampled)
this PR_1 (case insensitive) x 2,588,539 ops/sec ±0.23% (95 runs sampled)
this PR_2 x 2,666,781 ops/sec ±0.72% (91 runs sampled)
this PR_2 (case insensitive) x 2,609,032 ops/sec ±0.34% (95 runs sampled)
this PR_3 x 2,599,787 ops/sec ±0.60% (94 runs sampled)
this PR_3 (case insensitive) x 2,668,658 ops/sec ±0.45% (92 runs sampled)
this PR_4 x 2,557,027 ops/sec ±0.94% (94 runs sampled)
this PR_4 (case insensitive) x 2,635,131 ops/sec ±0.39% (95 runs sampled)
current_1 x 2,305,391 ops/sec ±0.86% (90 runs sampled)
current_1 (case insensitive) x 2,324,108 ops/sec ±0.37% (96 runs sampled)
current_2 x 2,406,469 ops/sec ±0.85% (93 runs sampled)
current_2 (case insensitive) x 2,419,272 ops/sec ±0.48% (93 runs sampled)
current_3 x 2,305,482 ops/sec ±1.07% (95 runs sampled)
current_3 (case insensitive) x 2,269,017 ops/sec ±3.06% (90 runs sampled)
current_4 x 2,503,994 ops/sec ±0.83% (95 runs sampled)
current_4 (case insensitive) x 2,446,695 ops/sec ±1.50% (90 runs sampled)
-------------------------------------------------
Fastest is this PR_3 (case insensitive),this PR_2

this PR_1 x 2,634,960 ops/sec ±0.42% (92 runs sampled)
this PR_1 (case insensitive) x 2,647,087 ops/sec ±0.40% (92 runs sampled)
this PR_2 x 2,685,761 ops/sec ±0.39% (88 runs sampled)
this PR_2 (case insensitive) x 2,607,277 ops/sec ±0.40% (93 runs sampled)
this PR_3 x 2,632,152 ops/sec ±0.33% (91 runs sampled)
this PR_3 (case insensitive) x 2,670,882 ops/sec ±0.30% (92 runs sampled)
this PR_4 x 2,631,017 ops/sec ±0.50% (92 runs sampled)
this PR_4 (case insensitive) x 2,680,430 ops/sec ±0.29% (91 runs sampled)
current_1 x 2,310,687 ops/sec ±0.38% (91 runs sampled)
current_1 (case insensitive) x 2,309,424 ops/sec ±0.38% (94 runs sampled)
current_2 x 2,450,252 ops/sec ±0.35% (92 runs sampled)
current_2 (case insensitive) x 2,464,966 ops/sec ±0.34% (92 runs sampled)
current_3 x 2,315,481 ops/sec ±0.44% (93 runs sampled)
current_3 (case insensitive) x 2,374,246 ops/sec ±0.35% (89 runs sampled)
current_4 x 2,533,402 ops/sec ±0.50% (92 runs sampled)
current_4 (case insensitive) x 2,477,094 ops/sec ±0.46% (92 runs sampled)
-------------------------------------------------
Fastest is this PR_2
```

Run 3:
```
this PR_1 x 2,594,635 ops/sec ±0.83% (90 runs sampled)
this PR_1 (case insensitive) x 2,585,571 ops/sec ±0.35% (95 runs sampled)
this PR_2 x 2,667,320 ops/sec ±0.75% (91 runs sampled)
this PR_2 (case insensitive) x 2,624,340 ops/sec ±0.39% (91 runs sampled)
this PR_3 x 2,622,451 ops/sec ±0.62% (92 runs sampled)
this PR_3 (case insensitive) x 2,604,862 ops/sec ±0.96% (92 runs sampled)
this PR_4 x 2,605,136 ops/sec ±0.81% (90 runs sampled)
this PR_4 (case insensitive) x 2,621,953 ops/sec ±0.42% (94 runs sampled)
current_1 x 2,368,028 ops/sec ±0.81% (94 runs sampled)
current_1 (case insensitive) x 2,309,547 ops/sec ±0.54% (95 runs sampled)
current_2 x 2,438,239 ops/sec ±0.77% (95 runs sampled)
current_2 (case insensitive) x 2,475,954 ops/sec ±0.33% (92 runs sampled)
current_3 x 2,333,289 ops/sec ±1.06% (94 runs sampled)
current_3 (case insensitive) x 2,350,654 ops/sec ±0.98% (92 runs sampled)
current_4 x 2,450,877 ops/sec ±1.00% (92 runs sampled)
current_4 (case insensitive) x 2,523,300 ops/sec ±0.42% (95 runs sampled)
-------------------------------------------------
Fastest is this PR_2

this PR_1 x 2,572,320 ops/sec ±0.44% (90 runs sampled)
this PR_1 (case insensitive) x 2,633,771 ops/sec ±0.39% (92 runs sampled)
this PR_2 x 2,660,766 ops/sec ±0.32% (91 runs sampled)
this PR_2 (case insensitive) x 2,618,535 ops/sec ±0.34% (88 runs sampled)
this PR_3 x 2,600,683 ops/sec ±0.49% (88 runs sampled)
this PR_3 (case insensitive) x 2,635,160 ops/sec ±0.36% (92 runs sampled)
this PR_4 x 2,647,031 ops/sec ±0.54% (93 runs sampled)
this PR_4 (case insensitive) x 2,620,891 ops/sec ±0.38% (95 runs sampled)
current_1 x 2,328,207 ops/sec ±0.35% (92 runs sampled)
current_1 (case insensitive) x 2,278,399 ops/sec ±0.38% (95 runs sampled)
current_2 x 2,481,722 ops/sec ±0.61% (88 runs sampled)
current_2 (case insensitive) x 2,497,186 ops/sec ±0.32% (92 runs sampled)
current_3 x 2,348,544 ops/sec ±0.39% (91 runs sampled)
current_3 (case insensitive) x 2,360,262 ops/sec ±0.40% (90 runs sampled)
current_4 x 2,485,966 ops/sec ±0.36% (94 runs sampled)
current_4 (case insensitive) x 2,542,923 ops/sec ±0.40% (90 runs sampled)
-------------------------------------------------
Fastest is this PR_2
```
I can do more runs if you want.

**Comment by RedSparr0w** [comment]
Thanks for that,
You guys okay with `PR_3 (case insensitive)`-`/^#([\da-f]{3}){1,2}$/i`?
From all the test i have done its usually close finish from PR_2 and PR_3

**Pull request merged**